File: content/examples/other/leetcode/problems/serialize-and-deserialize-binary-tree/example-1/solution.py
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def serialize_bfs(root):
    if root is None:
        return ""
    out = []
    q = deque([root])
    while q:
        node = q.popleft()
        if node is None:
            out.append("#")
            continue
        out.append(str(node.val))
        q.append(node.left)
        q.append(node.right)
    while out and out[-1] == "#":
        out.pop()
    return ",".join(out)


def deserialize_bfs(data):
    if not data:
        return None
    parts = data.split(",")
    root = TreeNode(int(parts[0]))
    q = deque([root])
    i = 1
    while q and i < len(parts):
        node = q.popleft()
        left = parts[i]
        i += 1
        if left != "#":
            node.left = TreeNode(int(left))
        q.append(node.left)
        if i >= len(parts):
            break
        right = parts[i]
        i += 1
        if right != "#":
            node.right = TreeNode(int(right))
        q.append(node.right)
    return root


if __name__ == "__main__":
    root = TreeNode(1, TreeNode(2), TreeNode(3, TreeNode(4), TreeNode(5)))
    s = serialize_bfs(root)
    print(s)
    print(serialize_bfs(deserialize_bfs(s)))

File: content/examples/other/leetcode/problems/serialize-and-deserialize-binary-tree/example-1/solution.js
class TreeNode {
  constructor(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function serializeBFS(root) {
  if (!root) return "";
  const out = [];
  const queue = [root];
  for (let qi = 0; qi < queue.length; qi += 1) {
    const node = queue[qi];
    if (!node) {
      out.push("#");
      continue;
    }
    out.push(String(node.val));
    queue.push(node.left);
    queue.push(node.right);
  }
  // trim trailing nulls
  while (out.length > 0 && out[out.length - 1] === "#") out.pop();
  return out.join(",");
}

function deserializeBFS(data) {
  if (!data) return null;
  const parts = data.split(",");
  const root = new TreeNode(Number(parts[0]));
  const queue = [root];
  let i = 1;
  for (let qi = 0; qi < queue.length && i < parts.length; qi += 1) {
    const node = queue[qi];
    const left = parts[i++];
    if (left !== undefined) {
      if (left !== "#") node.left = new TreeNode(Number(left));
      queue.push(node.left || null);
    }
    const right = parts[i++];
    if (right !== undefined) {
      if (right !== "#") node.right = new TreeNode(Number(right));
      queue.push(node.right || null);
    }
  }
  return root;
}

if (require.main === module) {
  const root = new TreeNode(1, new TreeNode(2), new TreeNode(3, new TreeNode(4), new TreeNode(5)));
  const s = serializeBFS(root);
  console.log(s);
  console.log(serializeBFS(deserializeBFS(s)));
}

module.exports = { TreeNode, serializeBFS, deserializeBFS };
